readium/shared/src/main/java/org/readium/r2/shared/resource/ZipContainer.kt

```diff
@@ -0,0 +1,188 @@
+
+        override suspend fun read(range: LongRange?): Try<ByteArray, Resource.Exception> =
+            try {
+                withContext(Dispatchers.IO) {
+                    val bytes =
+                        if (range == null)
+                            readFully()
+                        else
+                            readRange(range)
+                    Try.success(bytes)
+                }
+            } catch (e: IOException) {
+                Try.failure(Resource.Exception.Unavailable(e))
+            } catch (e: Exception) {
+                Try.failure(Resource.Exception.wrap(e))
+            }
+
+        private suspend fun readFully(): ByteArray =
+            withContext(Dispatchers.IO) {
+                archive.getInputStream(entry)
+                    .use {
+                        it.readFully()
+                    }
+            }
+
+        private fun readRange(range: LongRange): ByteArray =
+            stream(range.first).readRange(range)
+
+        /**
+         * Reading an entry in chunks (e.g. from the HTTP server) can be really slow if the entry
+         * is deflated in the archive, because we can't jump to an arbitrary offset in a deflated
+         * stream. This means that we need to read from the start of the entry for each chunk.
+         *
+         * To alleviate this issue, we cache a stream which will be reused as long as the chunks are
+         * requested in order.
+         *
+         * See this issue for more info: https://github.com/readium/r2-shared-kotlin/issues/129
+         */
+        private fun stream(fromIndex: Long): CountingInputStream {
+            // Reuse the current stream if it didn't exceed the requested index.
+            stream
+                ?.takeIf { it.count <= fromIndex }
+                ?.let { return it }
+
+            stream?.close()
+
+            return CountingInputStream(archive.getInputStream(entry))
+                .also { stream = it }
+        }
+
+        private var stream: CountingInputStream? = null
+
+        override suspend fun close() {
+            withContext(Dispatchers.IO) {
+                tryOrLog { stream?.close() }
+            }
+        }
+    }
+
+    override val file: File = source
+
+    override suspend fun name(): ResourceTry<String> =
+        ResourceTry.success(file.name)
+
+    override suspend fun entries(): List<Container.Entry> =
+        archive.entries().toList()
+            .filterNot { it.isDirectory }
+            .mapNotNull { Entry(it) }
+
+    override suspend fun entry(path: String): Container.Entry {
+        return archive.getEntry(path.removePrefix("/"))
+            ?.let { Entry(it) }
+            ?: FailureEntry(path)
+    }
+
+    override suspend fun close() {
+        tryOrLog {
+            withContext(Dispatchers.IO) {
+                archive.close()
+            }
+        }
+    }
+}
+
+class DefaultArchiveFactory : ArchiveFactory {
```

Comment:
I would move this to `Factories.kt`, or better a dedicated `ArchiveFactory.kt` file.